### ISSUE TITLE
Allow finer control over how GridLayer chooses which tiles to load and retain

### ISSUE THREAD

**Issue by irskep** [Leaflet/Leaflet#3087]
When panning around in a map with a `GridLayer`, tiles are only created once their rect enters the map's visible bounds. This means that as you pan around, you see tiles pop in constantly.

I'd like to have the ability to load additional tiles outside the view (for example, within a 200px margin) so that tiles are loaded before they enter the visible rect.

Possible fixes:

* Instead of calling `this._map.getBounds()` to determine the working set of tiles, write a simple wrapper `this._getBounds()` that can be overridden
* Take a `preloadMargin` argument (or something). Might be hard to document.

If @mourner has a preferred implementation, I'd be happy to submit a PR to implement this feature.  `GridLayer` has solved a lot of my problems and it would be cool to have it solve even more. :heart:

**Comment by irskep** [comment]
Ping for guidance

**Comment by IvanSanchez** [comment]
Possibly a duplicate of #512

**Comment by irskep** [comment]
Yeah, this is more a request for a plugin hook than an actual implementation of #512. Hence the wording of the title. I'd like to be able to plug in some kind of function that takes the current map bounds and returns a list of tiles to keep loaded.

@mourner said it could be implemented as a plugin, but from my experience with GridLayer's internals, I don't think it's very practical.

**Comment by mourner** [comment]
Yeah, I think we should make this easier in GridLayer for plugins. We'll be able to do that once we're done with events refactorings.

**Comment by mourner** [comment]
I think we're good for now after #3551
